### PR TITLE
BF: Allow passing kwargs in fit method, by moving parallelization kwargs elsewhere, including PEP 3102

### PR DESCRIPTION
**Pull request by arokem**
Replaces #3286

### CODE REVIEW

**Comment by arokem** [comment]
Pending merge of #3257. I can rebase once that's fixed up and merged.

File: dipy/reconst/shm.py [hunk 1]
```diff
@@ -399,10 +399,10 @@ def real_sh_descoteaux_from_index(m_values, l_values, theta, phi, *, legacy=True
 
     Y_l^m =
     \begin{cases}
-        \sqrt{2} * \Im(Y_l^m) \; if m > 0 \\
-        Y^0_l \; if m = 0 \\
-        \sqrt{2} * \Re(Y_l^m)  \; if m < 0 \\
-    \end{cases}
+        \\sqrt{2} * \\Im(Y_l^m) \\; if m > 0 \\
```
Comment (jhlegarreta) [inline dipy/reconst/shm.py:402]: Double backslashes should not be necessary if we use the raw string marker for the docstring. Would you mind doing the latter in a separate PR so that it can be merged independently and all doc-related PRs that have been opened lately can be rebased with less work? Thanks Ariel.

Comment (arokem) [reply dipy/reconst/shm.py:402]: ruff added these automatically. Do we need to configure it not to do that?

Comment (jhlegarreta) [reply dipy/reconst/shm.py:402]: :upside_down_face: Weird that it did that now and not before, and only on these files. I think using r involves less work, and the code is more readable.

File: dipy/reconst/multi_voxel.py [hunk 2]
```diff
@@ -71,13 +72,18 @@ def new_fit(self, data, mask=None, **kwargs):
...
+                if kk in kwargs:
+                    parallel_kwargs[kk] = kwargs[kk]
             fit_array[np.where(mask)] = np.concatenate(
                 (
                     paramap(
                         _parallel_fit_worker,
                         chunks,
                         func_args=[single_voxel_with_self],
-                        **kwargs,
+                        func_kwargs=kwargs,
```
Comment (samcoveney) [inline dipy/reconst/multi_voxel.py:85]: So can you please explain, that if I would like to pass in different keywords for each chunk, how to do this?

For example, I would need something like `weights=weights[ii : ii + vox_per_chunk]` such that each parallel worker is received a different weights `keyword`. I am still not sure how to achieve this here - would kwargs be replaced with a list of dictionaries, with each dictionary specifying the keywords for that chunk? Thanks

**Comment by skoudoro** [comment]
Hi @arokem,

#3257 has been merged. Can you rebase, fix all conflict and finalize this work?

Actually, I believe that after rebasing and fixing conflict, it will be ready to go. I will just try this PR locally before merging.

**Comment by arokem** [comment]
Sorry busy week. I'll do this first thing on Monday.

**Comment by arokem** [comment]
OK - just rebased this on master, and I believe it should be good to go (?)

File: dipy/reconst/dki.py [hunk 3]
```diff
@@ -466,7 +466,8 @@ def directional_diffusion(dt, V, *, min_diffusivity=0):
     return adc
 
 
-def directional_diffusion_variance(kt, V):
+@warning_for_keywords()
+def directional_diffusion_variance(kt, V, *, min_kurtosis=-3 / 7):
```
Comment (skoudoro) [inline dipy/reconst/dki.py:470]: problem from the rebase, to put it back to `def directional_diffusion_variance(kt, V):`

File: dipy/reconst/dti.py [hunk 4]
```diff
@@ -1574,7 +1574,8 @@ def _ols_fit_matrix(design_matrix):
 class _NllsHelper:
     r"""Class with member functions to return nlls error and derivative."""
 
-    def err_func(self, tensor, design_matrix, data, weighting=None, sigma=None):
+    @warning_for_keywords()
+    def err_func(self, tensor, design_matrix, data, *, weighting=None, sigma=None):
```
Comment (skoudoro) [inline dipy/reconst/dti.py:1578]: probleme with the rebase, to put it back to `def err_func(self, tensor, design_matrix, data, weighting=None, sigma=None):`

File: dipy/reconst/ivim.py [hunk 5]
```diff
@@ -285,6 +285,7 @@ def __init__(
 
         self.bounds = bounds or BOUNDS
 
+    @warning_for_keywords()
```
Comment (skoudoro) [inline dipy/reconst/ivim.py:288]: same as above

File: dipy/reconst/quick_squash.pyx [hunk 6]
```diff
@@ -15,7 +17,8 @@ SCALAR_TYPES = np.ScalarType
 
 @cython.boundscheck(False)
 @cython.wraparound(False)
-def quick_squash(obj_arr, mask=None, fill=0):
+@warning_for_keywords()
+def quick_squash(obj_arr, *, mask=None, fill=0):
```
Comment (skoudoro) [inline dipy/reconst/quick_squash.pyx:21]: same as above

File: dipy/reconst/recspeed.pyx [hunk 7]
```diff
@@ -33,9 +35,11 @@ cdef inline double* asdp(cnp.ndarray pt):
 
 @cython.boundscheck(False)
 @cython.wraparound(False)
+@warning_for_keywords()
 def remove_similar_vertices(
     cython.floating[:, :] vertices,
     double theta,
+    *,
     bint return_mapping=False,
```
Comment (skoudoro) [inline dipy/reconst/recspeed.pyx:43]: same as above

File: dipy/reconst/shm.py [hunk 8]
```diff
@@ -64,8 +64,8 @@ def forward_sdeconv_mat(r_rh, l_values):
         function. Each element ``rh[i]`` is associated with spherical harmonics
         of order ``2*i``.
     l_values : ndarray
-        The orders ($l$) of spherical harmonic function associated with each row
-        of the deconvolution matrix. Only even orders are allowed
+        The orders ($l$) of spherical harmonic function associated with each
+        row of the deconvolution matrix. Only even orders are allowed
```
Comment (skoudoro) [inline dipy/reconst/shm.py:68]: same as above

**Review by skoudoro — COMMENTED** [review]
Thank you for the rebase @arokem, 

See below some rebase issues.  I will test your PR in 2h and get back to you. I believe also it is ready to go but I just want to be certain.

File: dipy/reconst/dti.py (hunk 4 above)
Comment (arokem) [reply dipy/reconst/dti.py]: Done in 092b16850

File: dipy/reconst/ivim.py (hunk 5 above)
Comment (arokem) [reply dipy/reconst/ivim.py]: Done in 8a3a97756

File: dipy/reconst/quick_squash.pyx (hunk 6 above)
Comment (arokem) [reply dipy/reconst/quick_squash.pyx]: Done in 009dad8d7

File: dipy/reconst/recspeed.pyx (hunk 7 above)
Comment (arokem) [reply dipy/reconst/recspeed.pyx]: Done in eb40ef4ac

File: dipy/reconst/shm.py (hunk 8 above)
Comment (arokem) [reply dipy/reconst/shm.py]: Done in 59f3c2c0f

File: dipy/reconst/dki.py (hunk 3 above)
Comment (arokem) [reply dipy/reconst/dki.py]: Done in 1e5c01ff2

File: dipy/reconst/dki.py [hunk 9]
```diff
@@ -2336,6 +2336,7 @@ def mkt(self, *, min_kurtosis=-3.0 / 7, max_kurtosis=10):
             self.model_params, min_kurtosis=min_kurtosis, max_kurtosis=max_kurtosis
         )
 
+    @warning_for_keywords()
```
Comment (arokem) [inline dipy/reconst/dki.py:2339]: This is not currently on master, but seems like it does need to be here. Or am I wrong?

File: dipy/reconst/multi_voxel.py (hunk 2 above)
Comment (arokem) [reply dipy/reconst/multi_voxel.py:85]: Yes. I think that's right: a list of dictionaries.

File: dipy/reconst/dki.py (hunk 9 above)
Comment (skoudoro) [reply dipy/reconst/dki.py:2339]: we do not need it here because it was already a keyword only arguments function.

The decorator is here to warn user from old functions.  New functions like this one do not need it.

Comment (arokem) [reply dipy/reconst/dki.py]: OK. Removed in da97f4b47

File: dipy/reconst/quick_squash.pyx [hunk 10]
```diff
@@ -7,6 +7,8 @@ cimport cython
 
 import numpy as np
 
+from dipy.testing.decorators import warning_for_keywords
```
Comment (skoudoro) [inline dipy/reconst/quick_squash.pyx:10]: can you remove this import ?

Comment (arokem) [reply dipy/reconst/quick_squash.pyx]: I have no idea how this even got there, but it's gone as of fbe63ba66

File: dipy/reconst/recspeed.pyx [hunk 11]
```diff
@@ -10,6 +10,8 @@ cimport cython
 import numpy as np
 cimport numpy as cnp
 
+from dipy.testing.decorators import warning_for_keywords
```
Comment (skoudoro) [inline dipy/reconst/recspeed.pyx:13]: can you remove this import ?

Comment (arokem) [reply dipy/reconst/recspeed.pyx]: Done in d1c4e0919

File: dipy/reconst/tests/test_multi_voxel.py [hunk 12]
```diff
@@ -171,44 +180,63 @@ def predict(self, S0):
...
-    S0 = 100.0
-    npt.assert_equal(fit.predict(S0=S0), np.ones(many_voxels.shape) * S0)
-
-    # Test with parallelization (using the "serial" dummy engine)
-    fit = model.fit(many_voxels, engine="serial")
-
-    # If parallelization engines are installed use them to test:
-    if has_joblib:
-        fit = model.fit(many_voxels, engine="joblib")
+    for verbose in [True, False]:
```
Comment (skoudoro) [inline dipy/reconst/tests/test_multi_voxel.py:188]: Why do you iterate over verbose since you do not check the verbosity? curiosity

File: dipy/reconst/tests/test_multi_voxel.py [hunk 13]
```diff
@@ -171,44 +180,63 @@ def predict(self, S0):
...
-        fit = model.fit(many_voxels, engine="dask")
-        npt.assert_equal(fit.predict(S0=S0), np.ones(many_voxels.shape) * S0)
-
-    if has_ray:
-        fit = model.fit(many_voxels, engine="ray")
-        npt.assert_equal(fit.predict(S0=S0), np.ones(many_voxels.shape) * S0)
+    # Test with parallelization (using the "serial" dummy engine)
+    fit = model.fit(many_voxels, another_kwarg="foo", engine="serial")
+
+    for verbose in [True, False]:
```
Comment (skoudoro) [inline dipy/reconst/tests/test_multi_voxel.py:202]: Why do you iterate over verbose since you do not check the verbosity? curiosity

File: dipy/reconst/tests/test_multi_voxel.py [hunk 14]
```diff
@@ -171,44 +180,63 @@ def predict(self, S0):
...
+                npt.assert_equal(fit.predict(S0=S0), np.ones(many_voxels.shape) * S0)
+
+            if has_ray:
+                fit = model.fit(
+                    many_voxels,
+                    verbose=verbose,
+                    another_kwarg=another_kwarg,
+                    engine="ray",
+                )
+                npt.assert_equal(fit.predict(S0=S0), np.ones(many_voxels.shape) * S0)
```
Comment (skoudoro) [inline dipy/reconst/tests/test_multi_voxel.py:232]: is it relevant to check if it is faster than `serial` ?  small data so not sure it will be stable. Let me know what you think

Comment (arokem) [reply dipy/reconst/tests/test_multi_voxel.py:232]: I don't think we want to test this here. As you say, with this size data, it's not a given that it will always be faster, and may vary depending on other context: platform, hardware, etc.

File: dipy/reconst/tests/test_multi_voxel.py (hunk 13 above)
Comment (skoudoro) [reply dipy/reconst/tests/test_multi_voxel.py:202]: ok, got it ?

**Review by skoudoro — APPROVED** [review]
Looks good to me. Thank you for this great work @arokem. 

I think we still to have play more with this new feature, but for now, I will go ahead and merge this PR.

**Comment by samcoveney** [comment]
I'll see if I can make this work with #3170 now

**Comment by skoudoro** [comment]
> I'll see if I can make this work with https://github.com/dipy/dipy/pull/3170 now

Let us know if you need help for the rebase and this integration @samcoveney

**Comment by samcoveney** [comment]
Thanks, I will do. I'll probably have to just rewrite that part of the code entirely. Hopefully it'll work with this new PR

**Comment by samcoveney** [comment]
As things stand here (and I know this has been merged) this will currently not work with #3170 because the way the parallelization has been done in `dipy/utils/parallel.py` is not general enough.

Each parallel worker gets a chunk of the data in the mask, let's call it `data[ii : ii + vox_per_chunk]`, which consists of the data across a bunch of voxels. Each parallel worker can be coerced (with code changes to this PR...) to receive a dictionary of keywords in `multi_voxel.py`, which can include for example `weights=weights[ii : ii + vox_per_chunk]`, which are the weights needed for WLS fitting with `data[ii : ii + vox_per_chunk]`.

The problem here in `parallel.py` is that each worker is going to process each of the voxels in `data[ii : ii + vox_per_chunk]` one at a time (correct) - therefore every time the fitting function is called, it should be passing the corresponding weights for that voxel only _rather than the entire array_ `weights[ii : ii + vox_per_chunk]`. In other words, if processing `data[ii : ii + vox_per_chunk][0]`, the fitting function needs to be receiving `weights[ii : ii + vox_per_chunk][0]`

Sorry this is late, but I did keep asking for clarification of how this was really going to work. Any regression function with weights, needs to be receiving weights of the same size as the observations y, since of course the weights are applied to each observation. I feel that we could have tested this again #3170 before merging it.

At least you fixed the keywords bug I discovered, whereby keywords were not being passed whatsoever :)

So, what now?

**Comment by samcoveney** [comment]
Would something like this be okay in `dipy/reconst/multi_voxel.py` ? Still not sure of best solution working with what I've been given, but doing my best

```
def _parallel_fit_worker(vox_data, single_voxel_fit, **kwargs):
    """
    Works on a chunk of voxel data to create a list of
    single voxel fits.

    Parameters
    ----------
    vox_data : ndarray, shape (n_voxels, ...)
        The data to fit.

    single_voxel_fit : callable
        The fit function to use on each voxel.
    """
    vox_weights = kwargs.pop("weights")
    if type(vox_weights) is np.ndarray:
        return [single_voxel_fit(data, **(dict({"weights": weights}, **kwargs))) for data, weights in zip(vox_data, vox_weights)]
    else:
        return [single_voxel_fit(data, **kwargs) for data in vox_data]
```

**Comment by arokem** [comment]
Yeah - I think that would be fine. Would it work as a way to pass the
weights in for each voxel?
